executor/runtime/docker/docker.go

```diff
@@ -806,8 +739,76 @@ func prepareNetworkDriver(parentCtx context.Context, cfg Config, c *runtimeTypes
 	return nil
 }
 
+func (r *DockerRuntime) doSetupSSHdContainer(ctx context.Context, containerName *string) error { // nolint: gocyclo
+	tmpImageInfo, err := imageExists(ctx, r.client, r.cfg.ContainerSSHDImage)
+	if err != nil {
+		return err
+	}
+
+	if tmpImageInfo == nil {
+		err = pullWithRetries(ctx, r.metrics, r.client, r.cfg.ContainerSSHDImage, doDockerPull)
+		if err != nil {
+			return err
+		}
+		_, _, err = r.client.ImageInspectWithRaw(ctx, r.cfg.ContainerSSHDImage)
+		if err != nil {
+			return err
+		}
+	}
+
+	// so we replace @ with - to match Docker's image naming scheme, a la:
+	// [a-zA-Z0-9][a-zA-Z0-9_.-]
+	noDashes := strings.Replace(r.cfg.ContainerSSHDImage, ":", "-", -1)
+	noAts := strings.Replace(noDashes, "@", "-", -1)
+	noSlashes := strings.Replace(noAts, "/", "_", -1)
+	*containerName = "titus-sshd-" + noSlashes
+	_, err = r.client.ContainerInspect(ctx, *containerName)
+	if err == nil {
+		return nil
+	}
+
+	if !docker.IsErrNotFound(err) {
+		return err
+	}
+
+	cfg := &container.Config{
+		Hostname: "titus-sshd",
+		Volumes: map[string]struct{}{
+			"/titus/sshd": {},
+		},
+		Entrypoint: []string{"/bin/bash"},
+		Image:      r.cfg.ContainerSSHDImage,
+	}
+	hostConfig := &container.HostConfig{
+		NetworkMode: "none",
+	}
+	// Check if this container exists, if not create it.
+	// We don't check the error here, because there's no way
+	// to prevent us from accidentally calling this concurrently
+	_, tmpErr := r.client.ContainerCreate(ctx, cfg, hostConfig, nil, *containerName)
+	if tmpErr == nil {
+		return nil
+	}
+
+	// Do this with backoff
+	timer := time.NewTicker(100 * time.Millisecond)
```

Comment:
Does this ever stop?  If the `ContainerCreate` call above fails for a reason that's not related to another container creation in progress (so it will never be created), what's the behaviour?

Reply:
The timer itself will stop when we return from this function, see the `defer`. 

The loop itself can terminate under one of two conditions:
1) The container shows up and responds to inspect
2) If the context is cancelled. The context descends from:

```
// Prepare host state (pull image, create fs, create container, etc...) for the container
func (r *DockerRuntime) Prepare(parentCtx context.Context, c *runtimeTypes.Container, binds []string) error { // nolint: gocyclo
	var sshdContainerName string
	l := log.WithField("taskID", c.TaskID)
	l.WithField("prepareTimeout", r.dockerCfg.prepareTimeout).Info("Preparing container")

	ctx, cancel := context.WithTimeout(parentCtx, r.dockerCfg.prepareTimeout)
	defer cancel()
```

`	group, errGroupCtx := errgroup.WithContext(ctx)`

So, if the errgroup has an error, the context is cancelled. Alternatively, if the `func (r *Runner) prepareContainer(ctx context.Context, updateChan chan update) update` times out, it cancels the context.

Also, if we get a message from the master saying to die, it exits.